store/streaming/file/service.go

```diff
@@ -20,262 +22,150 @@ var _ baseapp.StreamingService = &StreamingService{}
 // It writes the received EndBlock request and response and the resulting state changes
 // out to a file as described in the above the naming schema
 func (fss *StreamingService) ListenEndBlock(ctx sdk.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) (rerr error) {
-	// generate the new file
-	dstFile, err := fss.openEndBlockFile()
-	if err != nil {
-		return err
-	}
-	defer func() {
-		cerr := dstFile.Close()
-		if rerr == nil {
-			rerr = cerr
-		}
-	}()
+	fss.blockMetadata.RequestEndBlock = &req
+	fss.blockMetadata.ResponseEndBlock = &res
+	return nil
+}
 
-	// write req to file
-	lengthPrefixedReqBytes, err := fss.codec.MarshalLengthPrefixed(&req)
-	if err != nil {
-		return err
+func (fss *StreamingService) ListenCommit(ctx sdk.Context, res abci.ResponseCommit) (err error) {
+	if fss.stopNodeOnErr {
+		defer func() {
+			if err != nil {
+				panic(err)
+			}
+		}()
 	}
-	if _, err = dstFile.Write(lengthPrefixedReqBytes); err != nil {
-		return err
+
+	fss.blockMetadata.ResponseCommit = &res
+
+	// write to target file atomically
+	metaFileName := fmt.Sprintf("block-%d-meta", fss.currentBlockNumber)
+	dataFileName := fmt.Sprintf("block-%d-data", fss.currentBlockNumber)
+	if fss.filePrefix != "" {
+		metaFileName = fmt.Sprintf("%s-%s", fss.filePrefix, metaFileName)
+		dataFileName = fmt.Sprintf("%s-%s", fss.filePrefix, dataFileName)
 	}
-	// write all state changes cached for this stage to file
-	fss.stateCacheLock.Lock()
-	for _, stateChange := range fss.stateCache {
-		if _, err = dstFile.Write(stateChange); err != nil {
-			fss.stateCache = nil
-			fss.stateCacheLock.Unlock()
+
+	if fss.outputMetadata {
+		bz, err := fss.codec.Marshal(&fss.blockMetadata)
+		if err != nil {
+			return err
+		}
+		if err := writeLengthPrefixedFile(path.Join(fss.writeDir, metaFileName), bz, fss.fsync); err != nil {
 			return err
 		}
 	}
-	// reset cache
-	fss.stateCache = nil
-	fss.stateCacheLock.Unlock()
-	// write res to file
-	lengthPrefixedResBytes, err := fss.codec.MarshalLengthPrefixed(&res)
-	if err != nil {
+
+	var buf bytes.Buffer
+	if err := fss.writeBlockData(&buf); err != nil {
 		return err
 	}
-	_, err = dstFile.Write(lengthPrefixedResBytes)
-	return err
+	return writeLengthPrefixedFile(path.Join(fss.writeDir, dataFileName), buf.Bytes(), fss.fsync)
 }
 
-func (fss *StreamingService) openEndBlockFile() (*os.File, error) {
-	fileName := fmt.Sprintf("block-%d-end", fss.currentBlockNumber)
-	if fss.filePrefix != "" {
-		fileName = fmt.Sprintf("%s-%s", fss.filePrefix, fileName)
+func (fss *StreamingService) writeBlockData(writer io.Writer) error {
+	for _, listener := range fss.storeListeners {
+		cache := listener.PopStateCache()
+		for i := range cache {
+			bz, err := fss.codec.MarshalLengthPrefixed(&cache[i])
+			if err != nil {
+				break
+			}
+			if _, err = writer.Write(bz); err != nil {
+				break
+			}
+		}
 	}
-	return os.OpenFile(filepath.Join(fss.writeDir, fileName), os.O_CREATE|os.O_WRONLY, 0o600)
+	return nil
```

Comment:
If fss.codec.MarshalLengthPrefixed or writer.Write return an error, where are those errors handled? Intuitively, it seems like they should cause the outer writeBlockData method to return an error, but it doesn't seem like that is the case?

Reply:
yeah, good catch, changed to return error now.